### PR TITLE
Add nullability to Localization

### PR DESCRIPTION
**Pull request by pranavkm**
* Changes to support nullable for ns2.0
* Change ref assembly project to use nullable context

### CODE REVIEW

File: eng/targets/CSharp.Common.targets [hunk 1]
```diff
@@ -28,6 +28,17 @@
     </ItemGroup>
   </Target>
 
+  <Choose>
```
Comment (pranavkm) [inline eng/targets/CSharp.Common.targets:31]: FYI @dotnet/aspnet-build

File: eng/targets/CSharp.Common.targets [hunk 2]
```diff
@@ -28,6 +28,17 @@
...
+    <When Condition="'$(Nullable)' != '' AND ('$(TargetFramework)' == 'netstandard2.0' OR '$(TargetFrameworkIdentifier)' == '.NETFramework')">
```
Comment (dougbu) [inline eng/targets/CSharp.Common.targets:32]: Could you add a comment explaining why .NET Core TFMs are excluded?

File: src/Localization/Abstractions/ref/Microsoft.Extensions.Localization.Abstractions.csproj [hunk 3]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <TargetFrameworks>netstandard2.0;$(DefaultNetCoreTargetFramework)</TargetFrameworks>
     <TargetFrameworks Condition="'$(DotNetBuildFromSource)' == 'true'">$(DefaultNetCoreTargetFramework)</TargetFrameworks>
+    <Nullable>annotations</Nullable>
```
Comment (dougbu) [inline src/Localization/Abstractions/ref/Microsoft.Extensions.Localization.Abstractions.csproj:6]: Ick, this gets ugly duplicating the src/ project's properties. Is it possible to add the (conditional) property in eng/targets/CSharp.ReferenceAssembly.props. I'm betting that would require a target in the src/ project that returns `$(Nullable)` and that's a _major_ hack as well as an `<MSBuild/>` explosion. I hope there's something we could do but don't have a good suggestion.

Another bad suggestion 😈 would be to load the src/ project into a `string` and see if it contains `<Nullable>annotations</Nullable>` or `<Nullable>enable</Nullable>`

File: src/Localization/Localization/src/ResourceManagerStringLocalizerFactory.cs [hunk 4]
```diff
@@ -228,17 +233,21 @@ protected virtual ResourceLocationAttribute GetResourceLocationAttribute(Assembl
...
             var rootNamespaceAttribute = GetRootNamespaceAttribute(assembly);
 
-            return rootNamespaceAttribute?.RootNamespace ??
-                new AssemblyName(assembly.FullName).Name;
+            if (rootNamespaceAttribute != null)
+            {
+                return rootNamespaceAttribute.RootNamespace;
+            }
+
+            return assembly.GetName().Name;
```
Comment (dougbu) [inline src/Localization/Localization/src/ResourceManagerStringLocalizerFactory.cs:250]: Just curious: Did the compiler complain about the previoius expression? If yes, what was it's concern?

File: src/Localization/Localization/test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs [hunk 5]
```diff
@@ -1,6 +1,7 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved. 
+// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+#nullable disable
```
Comment (dougbu) [inline src/Localization/Localization/test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs:4]: You're using nullable types. Why is this needed?

File: src/Shared/Nullable/NullableAttributes.cs [hunk 6]
```diff
@@ -0,0 +1,196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+// From https://github.com/dotnet/runtime/blob/ef72b95937703e485fdbbb75f3251fedfd1a0ef9/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/NullableAttributes.cs
+
+namespace System.Diagnostics.CodeAnalysis
+{
+    /// <summary>Specifies that null is allowed as an input even if the corresponding type disallows it.</summary>
+    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Parameter | AttributeTargets.Property, Inherited = false)]
+#if INTERNAL_NULLABLE_ATTRIBUTES
```
Comment (dougbu) [inline src/Shared/Nullable/NullableAttributes.cs:10]: Am I correct these attributes will be `public` when targeting .NET 5 and `internal` otherwise?

**Review by dougbu — APPROVED** [review]
Pending API review of course&hellip;

File: src/Localization/Abstractions/ref/Microsoft.Extensions.Localization.Abstractions.csproj (hunk 3 above)
Comment (pranavkm) [reply src/Localization/Abstractions/ref/Microsoft.Extensions.Localization.Abstractions.csproj:6]: Using `<Nullable>enable</Nullable>` in the ref assembly causes build errors because of all the of `throw null;` for instance, from the ref assembly below:

```C#
public string Value { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } }
```


Always using `annotations` seems like the right way unless we change the codegen to be null-safe.

File: src/Localization/Localization/src/ResourceManagerStringLocalizerFactory.cs (hunk 4 above)
Comment (pranavkm) [reply src/Localization/Localization/src/ResourceManagerStringLocalizerFactory.cs:250]: Both `FullName` and `Name` are nullable and `AssemblyName`'s ctor  does not like nulls. This had fewer null checks

File: src/Localization/Localization/test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs (hunk 5 above)
Comment (pranavkm) [reply src/Localization/Localization/test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs:4]: Some of these tests have dozens of null errors. I would like for new tests to be authored "correctly", but fixing these tests to either be correct or just litter it with null-forgiveness seems like a massive chore with low payout.

File: src/Shared/Nullable/NullableAttributes.cs (hunk 6 above)
Comment (pranavkm) [reply src/Shared/Nullable/NullableAttributes.cs:10]: This file doesn't get used in .NET Core since these types are already part of the runtime and we get them for free. They should be available as internal types in .NET Core. I copied this file as-is from dotnet/runtime with the hope that it would be easier to keep it updated as they add new annotations.

File: src/Localization/Localization/src/ResourceManagerStringLocalizerFactory.cs [hunk 7]
```diff
@@ -228,17 +233,21 @@ protected virtual ResourceLocationAttribute GetResourceLocationAttribute(Assembl
...
         }
 
-        private string GetRootNamespace(Assembly assembly)
+        private string? GetRootNamespace(Assembly assembly)
         {
             var rootNamespaceAttribute = GetRootNamespaceAttribute(assembly);
 
-            return rootNamespaceAttribute?.RootNamespace ??
-                new AssemblyName(assembly.FullName).Name;
+            if (rootNamespaceAttribute != null)
```
Comment (ryanbrandenburg) [inline src/Localization/Localization/src/ResourceManagerStringLocalizerFactory.cs:245]: I like it, much clearer.

**Review by ryanbrandenburg — APPROVED** [review]
Overall looks good.

File: src/Localization/Localization/src/Internal/AssemblyWrapper.cs [hunk 8]
```diff
@@ -25,8 +25,8 @@ public AssemblyWrapper(Assembly assembly)
 
         public Assembly Assembly { get; }
 
-        public virtual string FullName => Assembly.FullName;
+        public virtual string FullName => Assembly.FullName!;
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Localization/Localization/src/Internal/AssemblyWrapper.cs:28]: > Or does this actually mean "Assembly.FullName CLAIMS that it will return string?, but I know better and I'm telling you that that value will always be a non-null string"?

This

File: src/Localization/Localization/src/ResourceManagerStringLocalizer.cs [hunk 9]
```diff
@@ -171,7 +172,7 @@ protected IEnumerable<LocalizedString> GetAllStrings(bool includeParentCultures,
                 ? GetResourceNamesFromCultureHierarchy(culture)
                 : _resourceStringProvider.GetAllResourceStrings(culture, true);
 
-            foreach (var name in resourceNames)
+            foreach (var name in resourceNames ?? Enumerable.Empty<string>())
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Localization/Localization/src/ResourceManagerStringLocalizer.cs:175]: IMO, `if (resourceNames == null) { return; }`. Avoids allocating an `IEnumerable<string>`

Although this isn't a hot path so it doesn't really matter.

**Review by JamesNK — APPROVED** [review]

File: src/Localization/Localization/src/ResourceManagerStringLocalizer.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Localization/Localization/src/ResourceManagerStringLocalizer.cs:175]: AFAIK, these things get cached, so it's not saving much.
